### PR TITLE
Validation

### PR DESCRIPTION
**Pull request by trifonova-n**
Added validation dataset

### CODE REVIEW

File: gan/trainer.py [hunk 1]
```diff
@@ -94,12 +104,59 @@ def train_epoch(self, loader):
...
+            D_real, D_logit_real = self.discriminator(*sample)
+            noise = self.noise_sampler.sample_batch(loader.batch_size)
+            # empty tuple for not conditional gan
+            condition = noise[1:]
+
+            # G_sample is batch of images generated by generator using noise as input
+            G_sample = self.generator(*noise)
+            D_fake, D_logit_fake = self.discriminator(G_sample, *condition)
+
+            D_loss = discriminator_loss(D_logit_real, D_logit_fake)
```
Comment (Slonegg) [inline gan/trainer.py]: Shouldn't this block be reused? Invoking generator and discriminator.

File: gan/trainer.py [hunk 2]
```diff
@@ -94,12 +104,59 @@ def train_epoch(self, loader):
...
+
+            noise = self.noise_sampler.sample_batch(loader.batch_size)
+            # empty tuple for not conditional gan
+            condition = noise[1:]
+            G_sample = self.generator(*noise)
+            D_fake, D_logit_fake = self.discriminator(G_sample, *condition)
+            G_loss = generator_loss(D_logit_fake)
+            G_valid_loss += G_loss.data.cpu().numpy()
+            n_steps += 1
+            del sample, G_sample, D_real, D_logit_real, D_loss, D_fake, D_logit_fake, G_loss
```
Comment (Slonegg) [inline gan/trainer.py]: Is that necessary?

Comment (Slonegg) [reply gan/trainer.py]: Maybe wrapping into with statement will look nicer.

File: apps/visualize_network.py [hunk 3]
```diff
@@ -21,6 +24,9 @@
     visualizer.update_losses(2., 7., type='validation')
     visualizer.update_losses(4., 8., type='validation')
     visualizer.update_losses(6., 9., type='validation')
+    for batch, in loader:
+        visualizer.plot_batch(batch, ['img1', 'img2', 'img3', 'img4'])
+        break
```
Comment (Slonegg) [inline apps/visualize_network.py]: break here looks suspicious

**Review by Slonegg — APPROVED** [review]

File: gan/trainer.py (hunk 2 above)
Comment (trifonova-n) [reply gan/trainer.py]: Yes, that reduce memory consumption on GPU

File: apps/visualize_network.py (hunk 3 above)
Comment (trifonova-n) [reply apps/visualize_network.py]: That's just a test visualization script. I want to visualize 1 batch from dataloader to make sure visualization is correct

**Comment by Slonegg** [comment]
Then you can use `batch = next(iter(loader)) ` or something like that

File: gan/trainer.py (hunk 1 above)
Comment (trifonova-n) [reply gan/trainer.py]: Yes, this makes sense

File: gan/trainer.py (hunk 2 above)
Comment (trifonova-n) [reply gan/trainer.py]: How can I use with block for this?

**Pull request merged**